




library(ORFik)
library(Biostrings)
library(GenomicFeatures)

# Create a short DNAStringSet (example)
myDNA <- DNAStringSet(
  "AGAAGCCGCCCGGGTCTGCCCGGGGAAGGTATAGCCTTCGCCGGCTTCGAGGTAAGTATCCCCACCGCGCTTCGACCGCTAGGTCGAAGCGGGCCTCGGAGCCACCCCCTCGGGACATCGTTGTTGGAGGGGTTCCATGGCCTTTTTGGCACTCGCCCCGTTCTCTAACGCTCTCCCCGGGAGAAGAAGCAGATCGAAGCCGGTCGTGTCCCCGGGGAGCTCTTACCTCCGCAAGCCGAAGAAGGAGTGCCAAGAGCGGGTAAGCTTTCCAAGATGCGATCGATAGTCCTCGAAGGCTGGCTGGTCCAGTGAGCTGAAAGGCTCTCTAGTCCGCGATGCTACGATGGGTAAGCAACAGGTGCTTTATACTACTACGATGGAGTTTTGCCTTCCCCCTAGTGGGAGTGGCCAGCCCACACTATCGATTGTGATTGGCCGTTGATGATGGGCGGTGGGCGTGTAGCGGCTCTAGCCTATGGGGCCCGGTCCCGCGCTTTGCATTTGCATGCGCTTTTCGCCTCCCCCCCGCTCCAACCAATTAGAACCCGTGTGTCGTTTCTAATTTGCGTATGTCTCCTCCCAGGGAAGCGCGTCGCGCCAACGGGATGCCGAATAGCGCCTCTCATATGCATAAAGGTGAACGCCCCTGGACGCCATGACACCTCGATGCACATCTCATCTGCATGCGTCTCCTCCCCGGGAAGCGCGTCGCGCCAACGGGGTTCGTGATCGCGCCGCTCATATGCATAAAGACGAACGCCCCTGGACGCCATGACACTTCCTGGTAAATCTCATCTGCATACTGACGAGCTTGGGAGGAGCCGAGGGAGTGGGCTTCAAAAGTAATTTCAATAAAAATGGCGAGTGCGATATTTCCGACCGAAACGGAAATGATGTAAAAAAAGTGGGAGGGGGAGGGGGAAAGGTGGGCGTGAACGCGTCTCTGTATTTCCCGGTTGAATCTCATTAAAGTTATACCA"
); names(myDNA) <- 'NOIR'


myDNA <- DNAStringSet(
  "GGAAGCGCGTCGCGCCAACGGGATGCCGAATAGCGCCTCTCATATGCATAAAGGTGAACGCCCCTGGACGCCATGACACCTCGATGCACATCTCATCTGCATGCGTCTCCTCCCCGGGAAGCGCGTCGCGCCAACGGGGTTCGTGATCGCGCCGCTCATATGCATAAAGACGAACGCCCCTGGACGCCATGACACTTCCTGGTAAATCTCATCTGCATACTGACGAGCTTGGGAGGAGCCGAGGGAGTGGGCTTCAAAAGTAATTTCAATAAAAATGGCGAGTGCGATA"
); names(myDNA) <- 'NOIR_ex2'

# Find all possible ORFs using default start/stop codons (ATG, TAA/TAG/TGA)
orfs    <- findORFsFasta(myDNA,
                         startCodon = c("ATG"),        # e.g., consider GTG as a possible start
                         stopCodon  = c("TAA|TAG|TGA")  # default set
)

orfs_dt <- data.table(as.data.frame(orfs))


# Extract the sequences of the found ORFs
orf_sequences <- extractTranscriptSeqs(myDNA, orfs)

# Inspect the results
orfs
orf_sequences
